include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+  bool verifyDerived() const debug_function;
+
+  friend struct flatc::Type;
+};
+INTROSPECTION_NS(flatc, StructType, UUID, Name, Kind, Fields, Size);
+
+/// \brief A field of a union type in flatc, with position, qualified type, and
+/// name
+struct flatc::UnionField : flatc::AggregateField {
+  std::size_t Position;
+
+  UnionField(std::size_t P, const QualifiedType &QT, llvm::StringRef NameRef) :
+    AggregateField(QT, NameRef), Position(P) {}
+
+  UnionField(std::size_t P, QualifiedType &&QT, llvm::StringRef NameRef) :
+    AggregateField(std::move(QT), NameRef), Position(P) {}
+
+  UnionField(std::size_t P) : UnionField(P, {}, "") {}
+
+  UnionField() : UnionField(0) {}
+
+  ~UnionField() = default;
+
+  UnionField(const UnionField &) = default;
+  UnionField &operator=(const UnionField &) = default;
+
+  UnionField(UnionField &&) = default;
+  UnionField &operator=(UnionField &&) = default;
+
+  bool operator==(const UnionField &) const = default;
+};
+INTROSPECTION_NS(flatc, UnionField, FieldType, Name, Position);
+
+/// \brief KeyedObjectTraits for flatc::UnionField based on its position in the
+/// union
+template<>
+struct KeyedObjectTraits<flatc::UnionField> {
+
+  static std::size_t key(const flatc::UnionField &Val) { return Val.Position; }
+
+  static flatc::UnionField fromKey(const std::size_t &Position) {
+    return flatc::UnionField(Position);
+  }
+};
+
+/// \brief A union type in flatc. Uniont are actually typedefs of unnamed
+/// unions in C.
+struct flatc::UnionType : public flatc::Type {
+  static constexpr const char *Tag = "!flatc::UnionType";
+
+  SortedVector<UnionField> Fields;
+
+  UnionType(llvm::StringRef NameRef) :
+    Type(NameRef, TypeKind::Union), Fields() {}
+
+  UnionType() : UnionType("") {}
+
+  ~UnionType() = default;
+
+  UnionType(const UnionType &) = default;
+  UnionType &operator=(const UnionType &) = default;
+
+  UnionType(UnionType &&) = default;
+  UnionType &operator=(UnionType &&) = default;
+
+  static bool classof(const Type *T) { return T->getKind() == TypeKind::Union; }
+
+private:
+  // Kind must be TypeKind::Union, Fields must not be empty, all Fields must
+  // verify
+  bool verifyDerived() const debug_function;
+
+  friend struct flatc::Type;
+};
+INTROSPECTION_NS(flatc, UnionType, UUID, Name, Kind, Fields);
+
+/// \brief An argument type in a function pointer type in flatc.
+struct flatc::ArgumentType {
+  std::size_t Pos;
+  QualifiedType Type;
+
+  ArgumentType(std::size_t P, const QualifiedType &QT) : Pos(P), Type(QT) {}
+  ArgumentType(std::size_t P, QualifiedType &&QT) :
+    Pos(P), Type(std::move(QT)) {}
+
+  ArgumentType(std::size_t P) : ArgumentType(P, {}) {}
+
+  ArgumentType() : ArgumentType(0) {}
+
+  ~ArgumentType() = default;
+
+  ArgumentType(const ArgumentType &) = default;
+  ArgumentType &operator=(const ArgumentType &) = default;
+
+  ArgumentType(ArgumentType &&) = default;
+  ArgumentType &operator=(ArgumentType &&) = default;
```

Comment:
Is all of this really necessary?
Ideally in the model we should have more or less plain structs.

Reply:
Dropped